### PR TITLE
NBS-4814: [Disk Manager] implement VHD Reader

### PR DESCRIPTION
**Pull request by BarkovBG**

### CODE REVIEW

File: cloud/disk_manager/internal/pkg/dataplane/url/vhd/image_map_reader.go [hunk 1]
```diff
@@ -95,3 +138,70 @@ func (r *ImageMapReader) readHeader(ctx context.Context) error {
...
+		ctx,
+		r.header.TableOffset,
+		r.getBATSizeInBytes(),
+		binary.BigEndian,
+		&r.bat,
+	)
+}
+
+func (r *ImageMapReader) getBATSizeInBytes() uint64 {
+	return uint64(r.header.MaxTableEntries) * 32 / 8
```
Comment (jkuradobery) [inline cloud/disk_manager/internal/pkg/dataplane/url/vhd/image_map_reader.go:155]: BAT entry is 4 bytes long. Why * 32 / 8? And add a constant. The address space of the VHD seems to be 4 bytes unsigned int.

File: cloud/disk_manager/internal/pkg/dataplane/url/vhd/image_map_reader.go [hunk 2]
```diff
@@ -43,7 +52,41 @@ func (r *ImageMapReader) Read(ctx context.Context) ([]common.ImageMapItem, error
...
+		return []common.ImageMapItem{}, err
+	}
+
+	var items []common.ImageMapItem
+	var item common.ImageMapItem
+	var entry imageMapEntry
+
+	for i, _ := range r.bat {
+		offset := item.Start + item.Length
+		bytes := min(uint64(r.header.BlockSize), r.Size()-offset)
```
Comment (jkuradobery) [inline cloud/disk_manager/internal/pkg/dataplane/url/vhd/image_map_reader.go:66]: What does the `bytes` mean? Please, come up with another naming. Consider entrySizeBytes or something. It is a size.

File: cloud/disk_manager/internal/pkg/dataplane/url/vhd/common.go [hunk 3]
```diff
@@ -0,0 +1,29 @@
+package vhd
+
+import "github.com/ydb-platform/nbs/cloud/disk_manager/internal/pkg/dataplane/url/common"
+
+////////////////////////////////////////////////////////////////////////////////
+
+type imageMapEntry struct {
+	data   bool
```
Comment (jkuradobery) [inline cloud/disk_manager/internal/pkg/dataplane/url/vhd/common.go:8]: It should be hasData. data bool just sounds not right.

File: cloud/disk_manager/internal/pkg/dataplane/url/vhd/common.go [hunk 4]
```diff
@@ -0,0 +1,29 @@
+package vhd
+
+import "github.com/ydb-platform/nbs/cloud/disk_manager/internal/pkg/dataplane/url/common"
+
+////////////////////////////////////////////////////////////////////////////////
+
+type imageMapEntry struct {
```
Comment (jkuradobery) [inline cloud/disk_manager/internal/pkg/dataplane/url/vhd/common.go:7]: This structure duplicates the common.ImageMapItem, I do not think it is required.

Comment (BarkovBG) [reply cloud/disk_manager/internal/pkg/dataplane/url/vhd/common.go:7]: it is not a duplicate. i need it to implement mergeable and dumpToItem methods

code becomes easier 

we have the same structures in other formats:
https://github.com/ydb-platform/nbs/blob/bf37ab3e6fb00f5093ccd4d589f9875edd0e0abf/cloud/disk_manager/internal/pkg/dataplane/url/qcow2/common.go#L90

File: cloud/disk_manager/internal/pkg/dataplane/url/vhd/common.go (hunk 3 above)
Comment (BarkovBG) [reply cloud/disk_manager/internal/pkg/dataplane/url/vhd/common.go:8]: no, this field has the same naming with imageMapEntry from qemu-img map 
https://github.com/ydb-platform/nbs/blob/bf37ab3e6fb00f5093ccd4d589f9875edd0e0abf/cloud/disk_manager/internal/pkg/dataplane/url/common/image_map.go#L19

File: cloud/disk_manager/internal/pkg/dataplane/url/vhd/image_map_reader.go [hunk 5]
```diff
@@ -95,3 +138,70 @@ func (r *ImageMapReader) readHeader(ctx context.Context) error {
...
+}
+
+// GetBlockDataAddress returns the address of the 'data section' of a given
+// block. Address is the absolute byte offset of the 'data section'. A block
+// consists of 'block bitmap section' and 'data section'
+func (r *ImageMapReader) getBlockDataAddress(blockIndex uint32) uint64 {
+	return r.getBitmapAddress(blockIndex) + r.getSectorPaddedBitmapSizeInBytes()
+}
+
+func (r *ImageMapReader) readImageMapEntry(blockIndex uint32) imageMapEntry {
```
[the remark this answers is not in the document]
Comment (BarkovBG) [reply cloud/disk_manager/internal/pkg/dataplane/url/vhd/image_map_reader.go:195]: i want to implement the same structure of code with other formats 
https://github.com/search?q=repo%3Aydb-platform%2Fnbs+path%3Adisk_manager+readImageMapEntry&ref=opensearch&type=code

i think that code without this method will be more difficult for reading

File: cloud/disk_manager/internal/pkg/dataplane/url/vhd/common.go (hunk 4 above)
Comment (jkuradobery) [reply cloud/disk_manager/internal/pkg/dataplane/url/vhd/common.go:7]: I will look into the merge mechanism further.

File: cloud/disk_manager/internal/pkg/dataplane/url/vhd/image_map_reader.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (jkuradobery) [reply cloud/disk_manager/internal/pkg/dataplane/url/vhd/image_map_reader.go:195]: QCOW2 format is way more complicated then VHD.

File: cloud/disk_manager/internal/pkg/dataplane/url/vhd/common.go [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+	return e.hasData == other.hasData && e.offset == other.offset
+}
+
+func (e *imageMapEntry) dumpToItem(item *common.ImageMapItem) {
+	if e.hasData {
+		item.Zero = false
+		item.Data = true
+	} else {
+		item.Zero = true
+		item.Data = false
```
Comment (jkuradobery) [inline cloud/disk_manager/internal/pkg/dataplane/url/vhd/common.go:22]: item.data = e.hasData
item.zero = !e.hasData

File: cloud/disk_manager/internal/pkg/dataplane/url/vhd/image_map_reader.go [hunk 7]
```diff
@@ -43,7 +53,49 @@ func (r *ImageMapReader) Read(ctx context.Context) ([]common.ImageMapItem, error
...
+	err = r.readBAT(ctx)
+	if err != nil {
+		return []common.ImageMapItem{}, err
+	}
+
+	var items []common.ImageMapItem
+	var item common.ImageMapItem
+	var entry imageMapEntry
+
+	for i, _ := range r.bat {
```
Comment (jkuradobery) [inline cloud/disk_manager/internal/pkg/dataplane/url/vhd/image_map_reader.go:65]: The point of my proposal was to use r.bat iterator to access the address directly, without the necessity to use the index. Using the index if you have an iterator just looks ugly, despite the fact that I understand that the index is within the array range.

**Review by leftmain — DISMISSED** [review]

File: cloud/disk_manager/internal/pkg/dataplane/url/vhd/image_map_reader.go [hunk 8]
```diff
@@ -95,3 +147,56 @@ func (r *ImageMapReader) readHeader(ctx context.Context) error {
...
+	bitmapSizeInBytes := float64(r.getBitmapSizeInBytes())
+	return uint64(
+		math.Ceil(bitmapSizeInBytes/sectorSizeInBytes) * sectorSizeInBytes,
+	)
+}
+
+// GetBitmapAddress returns the address of the 'block bitmap section' of a given
+// block. Address is the absolute byte offset of the 'block bitmap section'. A
+// block consists of 'block bitmap section' and 'data section'
+func (r *ImageMapReader) getBitmapAddress(blockIndex uint32) uint64 {
```
Comment (jkuradobery) [inline cloud/disk_manager/internal/pkg/dataplane/url/vhd/image_map_reader.go:193]: getBitmapAddress should take an offset instead of the blockIndex.

**Review by jkuradobery — APPROVED** [review]

**Review by EvgeniyKozev — APPROVED** [review]

**Pull request merged**